src/webui/api/torrentscontroller.cpp

```diff
@@ -1622,3 +1792,246 @@ void TorrentsController::setSSLParametersAction()
+            const BitTorrent::InfoHash infoHash = torrentDescr.infoHash();
+
+            const QString &fileName = it.key();
+            m_torrentSource.insert(FILE_SOURCE_PREFIX + fileName, infoHash);
+            m_torrentMetadata.insert(infoHash, torrentDescr);
+
+            result.insert(fileName, serializeTorrentInfo(torrentDescr));
+            continue;
+        }
+
+        throw APIError(APIErrorType::BadData, tr("'%1' is not a valid torrent file.").arg(it.key()));
+    }
+
+    setResult(result);
+}
+
+void TorrentsController::saveMetadataAction()
+{
+    requireParams({u"source"_s});
+
+    const QString sourceParam = params()[u"source"_s].trimmed();
+    if (sourceParam.isEmpty())
+        throw APIError(APIErrorType::BadParams, tr("Must specify URI or hash"));
+
+    const QString source = QUrl::fromPercentEncoding(sourceParam.toLatin1());
+    if (const auto iter = m_torrentSource.constFind(source); iter != m_torrentSource.constEnd())
+    {
+        const BitTorrent::InfoHash infoHash = iter.value();
+        if (isMetadataDownloaded(infoHash))
+        {
+            const BitTorrent::TorrentDescriptor torrentDescr = m_torrentMetadata[infoHash];
+            const nonstd::expected<QByteArray, QString> result = torrentDescr.saveToBuffer();
+            if (!result)
+                throw APIError(APIErrorType::Conflict, tr("Unable to export torrent metadata. Error: %1").arg(result.error()));
+
+            setResult(result.value(), Http::CONTENT_TYPE_TORRENT_FILE, (infoHash.toTorrentID().toString() + u".torrent"));
+            return;
+        }
+
+        throw APIError(APIErrorType::Conflict, tr("Metadata is not yet available"));
+    }
+
+    throw APIError(APIErrorType::NotFound);
+}
+
+void TorrentsController::onDownloadFinished(const Net::DownloadResult &result)
+{
+    const QString source = result.url;
+
+    switch (result.status)
+    {
+    case Net::DownloadStatus::Success:
+        qDebug("Received torrent from %s", qUtf8Printable(source));
+
+        // use the info directly from the .torrent file
+        if (const auto loadResult = BitTorrent::TorrentDescriptor::load(result.data))
+        {
+            const BitTorrent::TorrentDescriptor &torrentDescr = loadResult.value();
+            const BitTorrent::InfoHash infoHash = torrentDescr.infoHash();
+
+            m_torrentSource.insert(source, infoHash);
+            m_torrentMetadata.insert(infoHash, torrentDescr);
+
+            return;
+        }
+
+        qDebug("Unable to parse torrent from %s", qUtf8Printable(source));
+        m_torrentSource.remove(source);
+        break;
+    case Net::DownloadStatus::RedirectedToMagnet:
+        if (const auto parseResult = BitTorrent::TorrentDescriptor::parse(result.magnetURI))
+        {
+            const BitTorrent::TorrentDescriptor &torrentDescr = parseResult.value();
+            const BitTorrent::InfoHash infoHash = torrentDescr.infoHash();
+
+            if (!m_torrentSource.contains(source))
```

Comment:
How can this be possible? You added it before downloading.

Reply:
Good catch, it's not possible. This is the code path that has been the toughest to test, but I'll see if I can set something up.